scylla/tests/retries.rs

```diff
@@ -0,0 +1,284 @@
+    for (real_addr, proxy_addr) in [
+        (real1_addr, proxy1_addr),
+        (real2_addr, proxy2_addr),
+        (real3_addr, proxy3_addr),
+    ] {
+        translation_map.insert(real_addr, proxy_addr);
+    }
+
+    let running_proxy = test(
+        [proxy1_uri, proxy2_uri, proxy3_uri],
+        translation_map,
+        running_proxy,
+    )
+    .await;
+
+    running_proxy.finish().await
+}
+
+#[tokio::test]
+#[ntest::timeout(5000)]
+async fn speculative_execution_is_fired() {
+    const TIMEOUT_PER_REQUEST: Duration = Duration::from_millis(200);
+    let res = test_with_3_node_cluster(217, |proxy_uris, translation_map, mut running_proxy| async move {
+        // DB preparation phase
+        let session: Session = SessionBuilder::new()
+            .known_node(proxy_uris[0].as_str())
+            .speculative_execution(Arc::new(SimpleSpeculativeExecutionPolicy {
+                max_retry_count: 2,
+                retry_interval: Duration::from_millis(10),
+            }))
+            .retry_policy(Box::new(FallthroughRetryPolicy))
+            .address_translator(Arc::new(translation_map))
+            .build()
+            .await
+            .unwrap();
+
+        session
+            .query("DROP KEYSPACE IF EXISTS ks2", &[])
+            .await
+            .unwrap();
+        session.query("CREATE KEYSPACE IF NOT EXISTS ks2 WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 3}", &[]).await.unwrap();
+        session
+            .query("CREATE TABLE ks2.t (a int primary key)", &[])
+            .await
+            .unwrap();
+
+        let mut q = Query::from("INSERT INTO ks2.t (a) VALUES (?)");
+        q.set_is_idempotent(true); // this is to allow speculative execution to fire
+
+        let delay_rule = RequestRule(
+            Condition::RequestOpcode(RequestOpcode::Query)
+                .and(Condition::BodyContainsCaseSensitive(Box::new(*b"ks2.t"))),
+            RequestReaction::drop_frame(),
+        );
+
+        info!("--------------------- BEGINNING main test part ----------------");
+
+        info!("--------------------- first query - no rules  ----------------");
+        // first run before any rules
+        tokio::select! {
+            res = session.query(q.clone(), (3,)) => res.unwrap(),
+            _ = tokio::time::sleep(TIMEOUT_PER_REQUEST) => panic!("Did not receive response in time."),
+        };
+
+        info!("--------------------- second query - 0 and 2 nodes not responding  ----------------");
+        running_proxy.running_nodes[0]
+            .change_request_rules(Some(vec![delay_rule.clone()]))
+            .await;
+        running_proxy.running_nodes[2]
+            .change_request_rules(Some(vec![delay_rule.clone()]))
+            .await;
+
+            tokio::select! {
+                res = session.query(q.clone(), (2,)) => res.unwrap(),
+                _ = tokio::time::sleep(TIMEOUT_PER_REQUEST) => panic!("Did not receive response in time."),
+            };
```

Comment:
Relying on timing in tests is a bad thing. You cannot reliably check that something didn't happen just by timing out - there is always a possibility that the thing you are waiting is executing really slow.

That said, I don't see a better way to test it - after all, you do want to check here that the response does not come back. I would increase the timeout, though, due to the slowness of the machines - let's increase it to 1 second.